File: src/generators/skill.ts
import type { Generator, GeneratorConfig, GeneratedFile, PromptQuestion } from './types.js'

function toSlug(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

function toTitleCase(value: string): string {
  return value
    .split(/[-_\s]+/)
    .filter(Boolean)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(' ')
}

function normalizeSteps(raw: string): string[] {
  const lines = raw
    .split(/\n+/)
    .map((line) => line.replace(/^\d+[.)]\s*/, '').trim())
    .filter(Boolean)

  if (lines.length > 0) {
    return lines
  }

  return ['Clarify scope', 'Execute minimal safe actions', 'Verify outcomes']
}

export class SkillGenerator implements Generator {
  readonly type = 'skill' as const

  getPromptQuestions(): PromptQuestion[] {
    return [
      {
        key: 'command',
        label: 'Command trigger (without leading slash)',
        type: 'text',
        required: true,
      },
      {
        key: 'steps',
        label: 'Workflow steps (newline or numbered list)',
        type: 'text',
        required: false,
      },
    ]
  }

  async generate(config: GeneratorConfig): Promise<GeneratedFile[]> {
    const slug = toSlug(config.name)
    const title = toTitleCase(slug || config.name)
    const command = String(config.answers?.command ?? (slug || 'command'))
    const goal = config.description ?? `Execute ${title.toLowerCase()} effectively.`
    const steps = normalizeSteps(String(config.answers?.steps ?? ''))

    const content = `# ${title} Skill

**Command:** /${command} [args]
**Goal:** ${goal}

---

## Workflow

${steps.map((step, index) => `${index + 1}. ${step}`).join('\n')}

## Principles

- Keep actions scoped and reversible.
- Prefer explicit assumptions over hidden behavior.
- Verify outputs before completion.

## Trace Protocol

For complex tasks, provide concise traces:

1. Thought
2. Action
3. Observation
4. Decision

## Output Format

\`\`\`markdown
## Skill Run: ${title}

### Steps Completed
- [step]: [status]

### Evidence
- [result or artifact]

### Follow-ups
- [if any]
\`\`\`
`

    return [
      {
        path: `library/skills/${slug || 'new-skill'}.md`,
        content,
      },
    ]
  }
}

File: packages/cli/internal/generator/skill.go
package generator

import (
	"fmt"
	"strings"

	"github.com/rluisb/lazyai/packages/cli/internal/types"
)

// SkillGenerator generates skill SKILL.md files.
// Ported from src/generators/skill.ts.
type SkillGenerator struct{}

func (g *SkillGenerator) Type() types.ArtifactType { return types.ArtifactTypeSkill }

func (g *SkillGenerator) GetPromptQuestions() []PromptQuestion {
	return []PromptQuestion{
		{
			Key:      "command",
			Label:    "Command trigger (without leading slash)",
			Type:     "text",
			Required: true,
		},
		{
			Key:      "steps",
			Label:    "Workflow steps (newline or numbered list)",
			Type:     "text",
			Required: false,
		},
	}
}

func (g *SkillGenerator) Generate(config GeneratorConfig) ([]GeneratedFile, error) {
	slug := ToSlug(config.Name)
	title := ToTitleCase(slug)
	if title == "" {
		title = config.Name
	}

	command := getAnswer(config.Answers, "command", slug)
	if command == "" {
		command = "command"
	}

	goal := config.Description
	if goal == "" {
		goal = fmt.Sprintf("Execute %s effectively.", strings.ToLower(title))
	}

	steps := normalizeSteps(getAnswer(config.Answers, "steps", ""))

	var stepLines []string
	for i, step := range steps {
		stepLines = append(stepLines, fmt.Sprintf("%d. %s", i+1, step))
	}

	content := fmt.Sprintf(`# %s Skill

**Command:** /%s [args]
**Goal:** %s

---

## Workflow

%s

## Principles

- Keep actions scoped and reversible.
- Prefer explicit assumptions over hidden behavior.
- Verify outputs before completion.

## Trace Protocol

For complex tasks, provide concise traces:

1. Thought
2. Action
3. Observation
4. Decision

## Output Format

`+"```"+`markdown
## Skill Run: %s

### Steps Completed
- [step]: [status]

### Evidence
- [result or artifact]

### Follow-ups
- [if any]
`+"```"+`
`,
		title,
		command,
		goal,
		strings.Join(stepLines, "\n"),
		title,
	)

	fileName := slug
	if fileName == "" {
		fileName = "new-skill"
	}

	return []GeneratedFile{
		{
			Path:    fmt.Sprintf("library/skills/%s.md", fileName),
			Content: content,
		},
	}, nil
}

// normalizeSteps parses a raw steps string into a list of steps.
func normalizeSteps(raw string) []string {
	var lines []string
	for _, line := range strings.Split(raw, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		// Remove leading number+period or number+paren.
		if len(line) > 2 {
			if line[0] >= '0' && line[0] <= '9' {
				if line[1] == '.' || line[1] == ')' {
					line = strings.TrimSpace(line[2:])
				} else if len(line) > 3 && (line[1] >= '0' && line[1] <= '9') && (line[2] == '.' || line[2] == ')') {
					line = strings.TrimSpace(line[3:])
				}
			}
		}
		if line != "" {
			lines = append(lines, line)
		}
	}

	if len(lines) > 0 {
		return lines
	}

	return []string{"Clarify scope", "Execute minimal safe actions", "Verify outcomes"}
}
